13-i-can-has-extension.md

```diff
@@ -0,0 +1,191 @@
+for extension messages.
+These message types are _multiplexed_,
+meaning that all extensions share the same message type numbers.
+Messages for a specific extension are called _extension-specific_
+messages.
+
+When a node receives a message of type 49154 to 65535,
+that message will have a meaning and format
+according to the "currently enabled extension" of the receiver.
+
+If the "currently enabled extension" is "no extension",
+then message types 49154 to 65535 will result in undefined behavior.
+
+Each node has its own concept of "currently enabled extension",
+which it uses for received messages.
+The `i_can_has_extension` message is used by the sender
+to change the "currently enabled extension" of the receiver.
+
+## Requirements
+
+Before sending an extension-specific message to the peer,
+the sending node MUST ensure that
+the "currently enabled extension" of the receiver is the extension
+the sending node intends to use.
+
+A sending node MAY send `i_can_has_extension` before every
+extension-specific message in order to ensure the above requirement.
+
+## Rationale
+
+By multiplexing, we allow a node to support multiple experimental
+features in parallel.
+Although messages are sent under one extension at a time only,
+the peer may switch between extensions at any time.
+
+Sending nodes can send `i_can_has_extension` before every
+extension-specific message so as to simplify implementations.
+Implementations can simply always send this before every
+extension-specific message so that they do not have to
+keep track of the state of the peer.
+
+Nodes supporting only a single extension can always just recognize
+the extension messages regardless of the purported
+"currently enabled extension".
+This is the reason for why "no extension" allows the
+multiplexed message types to result in undefined behavior:
+nodes that support only a single extension can simply not
+multiplex the message types and recognize them at all times,
+for simplicity.
+
+Both nodes over a connection have their own separate concepts of
+"currently enabled extension", rather than having a single
+"currently enabled extension" for both nodes.
+This avoids issues where one node wants to change the "currently
+enabled extension",
+at the same time as the other node is sending a message
+for the current extension.
+Senders are responsible for setting the "currently enabled extension"
+of the receiver, and the receiver does not change this by itself.
+
+# Informing Extensions
+
+A node implementation informs its peer of supported extensions
+by sending `i_has_a_extensions` message as soon as the
+peer connection is established.
+
+## The `i_has_a_extensions` Message
+
+This message contains a set of extensions supported by a node.
+
+1. type: 49153 (`i_has_a_extensions`)
+2. data:
+   * [`2` : `num_extensions`]
+   * [`num_extensions * 20` : `extensions`]
+
+### Requirements
+
+The sender:
+  - MUST send this message as soon as the connection is
+    established.
```

Comment:
The intent is that the sender must send this as soon as possible, so that the recipient can know immediately what extensions the sender has.

However, it is also the intent, that this message is optional, so possibly better wording can be had here.